src/jacobi.c

```diff
@@ -0,0 +1,219 @@
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when multiplying the matrices
+ */
+void jacobiMatrixMultiply(real_number* firstMatrix, real_number* secondMatrix, vec_size size[3], real_number* output, vec_size transposeFirstMatrix) {
+    
+    vec_size m = size[0];
+    vec_size n = size[1];
+    vec_size p = size[2];
+    for (vec_size i = 0; i < m; ++i) {
+        for (vec_size j = 0; j < p; ++j) {
+            real_number sum = 0;
+            for (vec_size k = 0; k < n; ++k) {
+                sum += (transposeFirstMatrix ? firstMatrix[coordToIndex(k, i, n)] : firstMatrix[coordToIndex(i, k, n)]) * secondMatrix[coordToIndex(k, j, n)];
+            }
+            output[coordToIndex(i, j, p)] = sum;
+        }
+    }
+}
+
+/**
+ * Creates a jacobi rotation matrix (1 on the diagonal and the value of cos(angle) and sin(angle) at row p and q)
+ * @param row row where to put the c and s values
+ * @param col column where to put the c and s values
+ * @param output output matrix (must be square)
+ * @param size size of the output matrix
+*/
+void jacobiCreateRotationMatrix(real_number* input, vec_size row, vec_size col, real_number* output, vec_size size) {
+
+    real_number aqq = input[coordToIndex(col, col, size)];
+    real_number app = input[coordToIndex(row, row, size)];
+    real_number apq = input[coordToIndex(row, col, size)];
+
+    real_number tau = (aqq - app) / (2.0 * apq);
+    real_number t1 = -tau - sqrt(1 + tau * tau);
+    real_number t2 = -tau + sqrt(1 + tau * tau);
+    real_number t = real_abs(t1) > real_abs(t2) ? t1 : t2;
+
+    real_number c = 1.0 / (sqrt(1 + t * t));
+    real_number s = c * t;
+
+    vec_size cIndex1 = coordToIndex(row, row, size);
+    vec_size cIndex2 = coordToIndex(col, col, size);
+    vec_size sIndex1 = coordToIndex(row, col, size);
+    vec_size sIndex2 = coordToIndex(col, row, size);
+    vec_size sizeSquared = size * size;
+
+    for (vec_size i = 0; i < sizeSquared; ++i) {
+        if (i == cIndex1 || i == cIndex2) {  
+            output[i] = c;
+        } else if (i == sIndex1) {
+            output[i] = s;
+        } else if (i == sIndex2) {
+            output[i] = -s;
+        } else if (i / size == i % size) {
+            output[i] = 1; // Diagonal
+        } else {
+            output[i] = 0; // Off diagonal
+        }
+    }
+}
+
+
+/**
+ * Creates the identity matrix (1 on the diagonal and 0 on other coordinates)
+ * @param size size of the matrix
+ * @param output output matrix
+*/
+void jacobiCreateIdentityMatrix(vec_size size, real_number* output) {
+    for (vec_size i = 0; i < size * size; ++i) {
+        if (i / size == i % size) {
```

Comment:
Do you think we should create an inline method isDiagonal(index) since this code is used at multiple places?

Reply:
Yes good point !